lib/Shell.js

```diff
@@ -0,0 +1,239 @@
+const { EventEmitter } = require('events');
+const { spawn } = require('child_process');
+const os = require('os');
+const shortid = require('shortid');
+
+const PSCommand = require('./PSCommand');
+
+const {
+  isWin, isCriticalPSError, logger,
+  ShellStreamBuffer, shellSafeWrite,
+  getType, convertToPSOption,
+} = require('./utils');
+
+const {
+  PS_PROC_ERROR,
+  PS_ARG_MISS_ERROR,
+  PS_ARG_TYPE_ERROR,
+  PS_CMD_FAIL_ERROR,
+} = require('./errors');
+
+class Shell extends EventEmitter {
+  constructor(options = {}) {
+    super();
+
+    this.pid = -1;
+    this.streams = [];
+    this.commands = [];
+    this.history = [];
+    this.hadErrors = false;
+    this.invocationStateInfo = 'NotStarted';
+    this.verbose = false;
+
+    let psProc = !isWin ? 'pwsh' : 'powershell';
+    let psOpts = ['-NoLogo', '-NoExit', '-Command', '-'];
+
+    if(options.pwsh) {
+      psProc = 'pwsh';
+    }
+
+    if(options.pwshPrev && !options.pwsh) {
+      psProc = 'pwsh-preview';
+    }
+
+    if(process.env.NPS) {
+      psProc = process.env.NPS;
+    }
+
+    if(Object.hasOwnProperty.call(options, 'debugMsg')) {
+      logger.warn(`debugMsg will be deprecated soon. use verbose instead.\n
+      https://rannn505.gitbook.io/node-powershell/shell/initialize`);
+      this.verbose = options.debugMsg;
+    }
+
+    if(options.verbose) {
+      this.verbose = true;
+    }
+
+    Object.keys(options).forEach(optKey => {
+      if(['pwsh', 'pwshPrev', 'debugMsg', 'verbose', 'inputEncoding', 'outputEncoding'].includes(optKey)) {
+        return;
+      }
+      const optVal = options[optKey];
+      psOpts = [...convertToPSOption(optKey, optVal), ...psOpts];
+    });
+
+    const proc = spawn(`${psProc}${!isWin ? '' : '.exe'}`, psOpts, { stdio: 'pipe' });
+    this.pid = proc.pid;
+
+    // make sure the PS process start successfully
+    if(!this.pid) {
+      throw new PS_PROC_ERROR();
+    }
+    proc.once('error', () => {
+      throw new PS_PROC_ERROR();
```

Comment:
FYI this is very aggressive behavior and means on any error in the child process the whole node process will crash and whomever is using this package cannot recover no matter what.

Reply:
(sans drastic stuff like overriding process.on('error' on the node process)

Reply:
this.emit('error', new PS_PROC_ERROR())